### PR TITLE
sem-ver implemented

### PR DESCRIPTION
**Pull request by msohailhussain**
## Summary
- Semantic version added
- Greater than or equal (ge) and less than or equal (le) evaluator added

## Test plan
- All relevant test cases added

### CODE REVIEW

File: packages/optimizely-sdk/lib/utils/semantic_version/index.tests.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+describe('lib/utils/sematic_version', function() {
+  describe('APIs', function() {
+    describe('compareVersion', function() {
+      it('should return 0 if user version and target version are equal', function() {
+        const versions = [
+          ['2.0.1', '2.0.1'],
+          ['2.9.9-beta', '2.9.9-beta'],
+          ['2.1', '2.1.0'],
+          ['2', '2.12'],
+          ['2.9', '2.9.1']
```
Comment (mnoman09) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.tests.js:28]: Add this test where targetSV = 2.1.3 and userver= 2.1.3+build and it should return 0 see this [test](https://github.com/optimizely/java-sdk/blob/2ffd6d9b0c3ad3900a5d93df27f8783024a39f9d/core-api/src/test/java/com/optimizely/ab/config/audience/SemanticVersionTest.java#L193)

**Review by mnoman09 — CHANGES_REQUESTED** [review]
Overall looks good. Added some comments.

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js [hunk 2]
```diff
@@ -760,4 +760,483 @@ describe('lib/core/custom_attribute_condition_evaluator', function() {
       );
     });
   });
+  describe('less than or equal to match type', function() {
```
[the remark this answers is not in the document]
Comment (ozayr-zaviar) [reply packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js:763]: ge and le evaluators does not not support alpha beta. Alpha and beta are only for semver

File: packages/optimizely-sdk/lib/utils/semantic_version/index.tests.js (hunk 1 above)
Comment (ozayr-zaviar) [reply packages/optimizely-sdk/lib/utils/semantic_version/index.tests.js:28]: 2.1.3 > 2.1.3+build

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js [hunk 3]
```diff
@@ -760,4 +760,483 @@ describe('lib/core/custom_attribute_condition_evaluator', function() {
...
+    var leCondition = {
+      match: 'le',
```
[the remark this answers is not in the document]
Comment (ozayr-zaviar) [reply packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js:765]: No version in evaluators (gt, lt, ge, le, etc)

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.ts [hunk 4]
```diff
@@ -211,6 +233,53 @@ function greaterThanEvaluator(condition: Condition, userAttributes: UserAttribut
...
+ *                             null if the condition value isn't a number or the user attribute value isn't a
+ *                             number
+ */
+function greaterOrEqualThanEvaluator(condition: Condition, userAttributes: UserAttributes): boolean | null {
+  const conditionName = condition.name;
+  const userValue = userAttributes[conditionName];
+  const userValueType = typeof userValue;
+  const conditionValue = condition.value;
+
+  if (conditionValue === null || !isSafeInteger(conditionValue)) {
```
[the remark this answers is not in the document]
Comment (ozayr-zaviar) [reply packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.ts:252]: Consistency in previous written evaluators like greaterThanEvaluator, lessThanEvaluator etc will be lost.

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mnoman09) [reply packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js:763]: obviously that's what I meant to add few beta alpha cases for semantic versioning in this file too.

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnoman09) [reply packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.tests.js:765]: same as above. I know what gt lt GE etc is  I was talking about semantic version

File: packages/optimizely-sdk/lib/utils/semantic_version/index.tests.js (hunk 1 above)
Comment (mnoman09) [reply packages/optimizely-sdk/lib/utils/semantic_version/index.tests.js:28]: No this is a corner case that we covered in all other sdks. you can see java link I shared above.

**Comment by mnoman09** [comment]
lgtm

**Review by mnoman09 — APPROVED** [review]
lgtm

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjc1283) [reply packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.ts:252]: I agree with @mnoman09. If there is repeated logic for checking user inputs for numeric comparison conditions, let's use a shared helper function. If the logic is not exactly the same or it is awkward to share, don't worry about it.

File: packages/optimizely-sdk/lib/utils/enums/index.ts [hunk 5]
```diff
@@ -226,3 +226,11 @@ export const DATAFILE_VERSIONS = {
...
   V4: '4',
 };
+
+/*
+* Pre-Release and Build symbols
+*/
+export const VERSION_TYPE = {
+  IS_PRE_RELEASE: '-',
+  IS_BUILD: '+'
+}
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/enums/index.ts:236]: TypeScript has `enum` and `const enum` declarations. Some can be compiled into constant values leading to smaller code size. Please check the [documentation for enums](https://www.typescriptlang.org/docs/handbook/enums.html). Let's not use an object with long property names unless necessary, to avoid bloating the code.

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    
+    if (targetSuffix) {
+        targetVersionParts.push(targetSuffix)
+    }
+
+    return targetVersionParts
+      
+  }
+  
+  export function compareVersion(conditionsVersion: string, userProvidedVersion: string): number | null {
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:141]: Please add a documentation comment

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    const isPreReleaseInconditionsVersion = isPreReleaseVersion(conditionsVersion)
+    const isPreReleaseInuserProvidedVersion = isPreReleaseVersion(userProvidedVersion)
+    const isBuildInconditionsVersion = isBuildVersion(conditionsVersion)
+  
+    const userVersionParts = splitVersion(userProvidedVersion);
+    const conditionsVersionParts = splitVersion(conditionsVersion);
+  
+    if (!userVersionParts || !conditionsVersionParts)
+      return null;
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:150]: Style preference is to always use brackets with if statements.

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        }
+      }
+      else {
+        const userVersionPart = parseInt(userVersionParts[idx])
+        const conditionsVersionPart = parseInt(conditionsVersionParts[idx])
+        if (userVersionPart > conditionsVersionPart)
+            return 1;
+        else if (userVersionPart < conditionsVersionPart)
+            return -1;
+      }
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:172]: Please try to simplify the control flow, there are a lot of nested conditionals.

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+const logger = getLogger();
+
+/**
+  * Evaluate if provided string is number only
+  * @param  {unknown}  content
+  * @return {boolean}  true if the string is number only
+  *                     
+  */
+  function isNumber(content: string): boolean {
+    return content.match(/^[0-9]+$/) != null ? true : false;
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:29]: Is this equivalent?
```suggestion
    return /^\d+$/.test(content);
```
(see: [Regular Expression docs for JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes))

Also: I suggest declaring regular expression once higher up in module scope, then reference here (`DIGIT_STRING_REGEXP.test(content)`).

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  }
+  
+  /**
+   * check if there is any white spaces " " in version
+   * @param  {unknown}  version
+   * @return {boolean}  true if the version contains " "
+   *                     
+   */
+  function hasWhiteSpaces(version: string): boolean {
+    return version.includes(' ');
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:81]: What about other space characters (tab, newline, etc.)? Should we use space regexp character class?
```suggestion
    return /\s/.test(version);
```
- Minor concern: `String.prototype.includes` is not supported in very old JS environments.
- Is there any performance difference between `String.prototype.includes` and regular expression `test` method?

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    
+    // check that version shouldn't have white space
+    if (hasWhiteSpaces(version)) {
+      logger.warn(LOG_MESSAGES.UNKNOWN_MATCH_TYPE, MODULE_NAME, version);
+      return null;
+    }
+    //check for pre release e.g. 1.0.0-alpha where 'alpha' is a pre release
+    //otherwise check for build e.g. 1.0.0+001 where 001 is a build metadata
+    if (isPreReleaseVersion(version)) {
+      targetPrefix = version.substring(0, version.indexOf(VERSION_TYPE.IS_PRE_RELEASE));
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:102]: `IS_PRE_RELEASE` sounds like a boolean. I prefer `PRE_RELEASE_VERSION_DELIMITER`.

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+      targetSuffix = version.substring(version.indexOf(VERSION_TYPE.IS_PRE_RELEASE) + 1);
+    }
+    else if (isBuildVersion(version)) {
+        targetPrefix = version.substring(0, version.indexOf(VERSION_TYPE.IS_BUILD));
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:106]: `BUILD_VERSION_DELIMITER`

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+   */
+  function hasWhiteSpaces(version: string): boolean {
+    return version.includes(' ');
+  }
+  
+  /**
+   * split version in parts
+   * @param  {unknown}  version
+   * @return {boolean}  The array of version split into smaller parts i.e major, minor, patch etc
+   *                    null if given version is in invalid format
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:88]: Would consumer code be clearer if we had an interface for semantic version components?
```ts
interface SemanticVersionComponents {
  major: number;
  minor: number;
  patch: number;
  preRelease?: number
  build?: number
}
```

Then:
```ts
function splitVersion(version: string): SemanticVersionComponents | null {
```
